src/Compiler/Driver/CompilerDiagnostics.fs

```diff
@@ -345,6 +349,10 @@ type Exception with
 #endif
         | ErrorsFromAddingSubsumptionConstraint(_, _, _, _, _, ContextInfo.DowncastUsedInsteadOfUpcast _, _) ->
             fst (FSComp.SR.considerUpcast ("", ""))
+        | ConstraintSolverNullnessWarningEquivWithTypes _ -> 3261
```

Comment:
This is worth a separate thorough design (design of the implementation within the compiler) reflecting the expressive power of the language and all the places where things like that can occur.

Having a decision tree for all possible constructs deep enough and surrounding context wide enough to be able to correctly suggest `xs was inferred` .. `or remove null from the collection` is a non-trivial thing to do if it should be done generically for all similar situations and not just special cased for this isolated scenario.

I think this deserves its own feature (not a small one) and dedicated allocation.


Having a single place listing all encountered classes of issues with diagnostics will definitely help with generalizing the design, it's good that you tested them and have some scenarios already!

Reply:
I created a few issues for diagnostics, I think they could be separate since will be most likely handled separately as well. They can be found by using area-nullness + area-diagnostics labels.